### ISSUE TITLE
Hash doesn't match that for other equal AbstractStrings

### ISSUE THREAD

**Issue by oxinabox** [xiaodaigh/ShortStrings.jl#21]
@scottpjones pointed this out on Slack.
and today it broke something i needed.

```julia
julia> ShortString3("o2") in Set(["o1", "o2"])
false
```

**Comment by xiaodaigh** [comment]
I see. Seems like it would be useful, wouldn't that make hash slower? Since there is a hash requirement?

`ShortString3("o2") in Set(["o1", "o2"])` so `in` is using hash? Shouldn't it try to do comparison without hash?

```
1.0 in Set([1,2]) # return true
```

I wonder if there is a way to make it do equality checks without hash.

**Comment by xiaodaigh** [comment]
fixed by #16

**Comment by oxinabox** [comment]
> I see. Seems like it would be useful, wouldn't that make hash slower? Since there is a hash requirement?

Probably. Lots of things can be fast but wrong thought.
But it is a requirment that equal things have equal hashes.
(but not the reverse).


> `ShortString3("o2") in Set(["o1", "o2"])` so `in` is using hash? Shouldn't it try to do comparison without hash?

The way `Set`s and `Dicts` in julia work is (roughly).
We have a `buckets = Vector{Vector{K}}` with some size.
Then to check if some object `x` is in the set we check:
```
the_bucket = buckets[mod1(hash(x),end)]  #everything in this bucket has same hash, modulo the number of buckets
for item in the_bucket
    isequal(item, x) && return true
end
return false
```


> ```
> 1.0 in Set([1,2]) # return true
> ```

Through some very clever process julia manages to hash floats and ints to be equal, if the values are equal.
```julia
julia> hash(1)
0x02011ce34bce797f

julia> hash(1.0)
0x02011ce34bce797f
```

> I wonder if there is a way to make it do equality checks without hash.

Sure, but you need to use a different collection type, like a `Vector` or `Tuple` will do brute-force equality checking.
A `AcceleratedVector` based on sorting will use `lessthan`, as will a `DataStructures.SortedSet` (which is a tree).
A `Trie` will use equality of elements.
I benchmarked SortedSet and it is slower for my use cases than paying the cost of converting to a string to hash
and I suspect much slower than if we could hash with MurmerHash3
